docs/_posts/2023-09-12-java-jni-benchmarks.markdown

```diff
@@ -0,0 +1,291 @@
+And RocksJava API implementation seamlessly continues to use the `std::string`-based `get()`
+
+Let's examine what happens when get() is called from Java
+
+```cpp
+jint Java_org_rocksdb_RocksDB_get__JJ_3BII_3BIIJ(
+   JNIEnv* env, jobject, jlong jdb_handle, jlong jropt_handle, jbyteArray jkey,
+   jint jkey_off, jint jkey_len, jbyteArray jval, jint jval_off, jint jval_len,
+   jlong jcf_handle)
+```
+
+ 1. Create an empty `std::string value`
+ 2. Call `DB::Get()` using the `std::string` variant
+ 3. Copy the resultant `std::string` into Java, using the JNI `SetByteArrayRegion()` method
+ 
+So stage (3) costs us a copy into Java. It's mostly unavoidable that there will be at least the one copy from a C++ buffer into a Java buffer.
+
+But what does stage 2 do ?
+
+ * Create a `PinnableSlice(std::string&)` which uses the value as the slice's backing buffer.
+ * Call `DB::Get()` using the PinnableSlice variant
+ * Work out if the slice has pinned data, in which case copy the pinned data into value and release it.
+ * ..or, if the slice has not pinned data, it is already in value (because we tried, but couldn't pin anything).
+
+So stage (2) costs us a copy into a `std::string`. But! It's just a naive `std::string` that we have copied a large buffer into. And in RocksDB, the buffer is or can be large, so an extra copy something we need to worry about.
+
+Luckily this is easy to fix. In the Java API (JNI) implementation:
+
+ 1. Create a PinnableSlice() which uses its own default backing buffer.
+ 2. Call `DB::Get()` using the `PinnableSlice` variant of the RocksDB API
+ 3. Copy the data indicated by the `PinnableSlice` straight into the Java output buffer using the JNI `SetByteArrayRegion()` method, then release the slice.
+ 4. Work out if the slice has successfully pinned data, in which case copy the pinned data straight into the Java output buffer using the JNI `SetByteArrayRegion()` method, then release the pin.
+ 5. ..or, if the slice has not pinned data, it is in the pinnable slice's default backing buffer. All that is left, is to copy it straight into the Java output buffer using the JNI SetByteArrayRegion() method.
+
+In the case where the `PinnableSlice` has succesfully pinned the data, this saves us the intermediate copy to the `std::string`. In the case where it hasn't, we still have the extra copy so the observed performance improvement depends on when the data can be pinned. Luckily, our benchmarking suggests that the pin is happening in a significant number of cases.
+
+On discussion with the RocksDB core team we understand that the core `PinnableSlice` optimization is most likely to succeed when pages are loaded from the buffer cache, rather than when they are in `memtable`. There is also a suggestion that with some coding effort, it might be possible to successfully pin in the buffer cache as well. This would likely improve the results for these benchmarks.
```

Comment:
I think you mean "it might be possible to successfully pin in the memtable as well." And "block cache" (RocksDB) instead of "buffer cache" (OS).

Reply:
tidied